### ISSUE TITLE
Exception in dune build when Unix domain socket in directory

### ISSUE THREAD

**Issue by verbosemode** [ocaml/dune#3124]
The **dune build** command raises the following exception when a Unix domain socket file is in the same directory:

Error: exception Sys_error("fail.sock: No such device or address")

## Reproduction

1.) Create a new directory

2.) Create a "dune' file with the following content
```
(executable
 (name unixsocket)
 (libraries unix))
```

3.) Create a "dune-project" file with the following content

`(lang dune 2.2)`

4.) Creating an empty .ml file without code.

`echo "" > unixsocket.ml`

5.) Create a Unix domain socket file (press ctrl+c to quit netcat)

`nc -Ul fail.sock`

6.) Run "dune build"

## Specifications

- Version of `dune` (output of `dune --version`): 2.2.0
- Version of `ocaml` (output of `ocamlc --version`): 4.09.0
- Operating system (distribution and version): Debian GNU/Linux 10 (buster)

## Additional information 

https://gist.github.com/verbosemode/7e7374309df5d6436a3400737fd8b3ad

**Comment by verbosemode** [comment]
I ran some more tests.

Creating a device file results in the same issue.

`sudo mknod mydevicefile c 0 0`

When I'm creating a FIFO with "mkfifo testfifo", dune build just gets stuck.

**Comment by diml** [comment]
Raising is unfortunate indeed. What behaviour would you expect? The best I can think of is making Dune ignore such files.

**Comment by Leonidas-from-XIV** [comment]
I think plainly ignoring such "special files" (except symlinks) might be the best idea, since it is unlikely that a build system can have a sensible interaction with e.g. a device file.

**Comment by verbosemode** [comment]
> Raising is unfortunate indeed. What behaviour would you expect? The best I can think of is making Dune ignore such files.

I agree with @Leonidas-from-XIV here. Ignoring non-regular files might be a good idea. A warning message and asking the user to clean up the directory would be too noisy and not so user friendly I guess.

**Comment by diml** [comment]
That seems fine to me
